docs/devel/api-conventions.md

```diff
@@ -370,6 +371,25 @@ All compatible Kubernetes APIs MUST support "name idempotency" and respond with
 
 Names generated by the system may be requested using `metadata.generateName`. GenerateName indicates that the name should be made unique by the server prior to persisting it. A non-empty value for the field indicates the name will be made unique (and the name returned to the client will be different than the name passed). The value of this field will be combined with a unique suffix on the server if the Name field has not been provided. The provided value must be valid within the rules for Name, and may be truncated by the length of the suffix required to make the value unique on the server. If this field is specified, and Name is not present, the server will NOT return a 409 if the generated name exists - instead, it will either return 201 Created or 504 with Reason `ServerTimeout` indicating a unique name could not be found in the time allotted, and the client should retry (optionally after the time indicated in the Retry-After header).
 
+## Optional vs Required
+
+Fields must be either optional or required.
+
+Optional fields have the following properties:
+
+- They have `omitempty` struct tag in Go.  This causes swagger documentation to reflect that the field is empty.
+- They are a pointer type in the Go definition (e.g. `bool *awesomeFlag`).  Using a pointer allows distinguishing unset from the default value.
```

Comment:
We've been inconsistent about this so far (e.g., spec and status). I think we would benefit from being more consistent, at least going forward. It may be too much work to fix existing types.

We've discussed doing this for a long, long time. For instance:
https://github.com/kubernetes/kubernetes/issues/2726#issuecomment-67681313

A pointer is at least required when we need to distinguish unset from the Go default value for the type.

Strings *maybe* don't have to be pointers in the case that the empty string is not a valid value.

I think I agree that arrays don't usually need to be pointers -- at least I can't think of any examples that would require it. The new label selector is a near miss, though, since no requirements matches everything but nil matches nothing.

POD types like int and bool probably should be pointers, as should most/all structs. Structs, in general, aren't omitted from encoder output even where omitempty is specified, which is messy.

With protobufs, we'd probably make every optional field a message in order to get the right behavior. Making every optional field a pointer is not worse than that.

cc @lavalamp